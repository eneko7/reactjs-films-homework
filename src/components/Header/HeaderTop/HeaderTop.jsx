import React from 'react';
import PropTypes from 'prop-types';
import style from './HeaderTop.scss';

class HeaderTop extends React.Component {
  constructor(props) {
    super(props);
    this.onChange = this.onChange.bind(this);
    this.searchFilms = this.searchFilms.bind(this);
  }

  onChange(e) {
    const { searchFilm, fetchFilmsPopular, saveSearchingWord } = this.props;
    if (e.target.value !== '') {
      searchFilm(e.target.value);
      saveSearchingWord(e.target.value);
    } else {
      fetchFilmsPopular();
    }
  }

  searchFilms() {
    const { searchFilm } = this.props;
    searchFilm(document.getElementById('search_input').value);
  }

  render() {
    const { searchedWord } = this.props;
    return (
      <div className={style.header_top}>
        <span className={style.header_top_span}><a href="/" className={style.header_top_span_a}>FILMS</a></span>
        <div className={style.header_top_search}>
          <input
            type="text"
            id="search_input"
            className={style.header_top_search_input}
            onChange={this.onChange}
            placeholder="Searh..."
            value={searchedWord}
          />
          <button className={style.header_top_search_label} onClick={this.searchFilms} type="button" />
        </div>
      </div>
    );
  }
}
HeaderTop.propTypes = {
  fetchFilmsPopular: PropTypes.func.isRequired,
  searchFilm: PropTypes.func.isRequired,
  saveSearchingWord: PropTypes.func.isRequired,
  searchedWord: PropTypes.string.isRequired,
};
export default HeaderTop;
