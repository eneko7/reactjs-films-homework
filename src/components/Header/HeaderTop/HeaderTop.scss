.header_top {
  .header_top_span {
    display: block;
    float: left;
    font-size: 35px;
    letter-spacing: 1px;
    line-height: 45px;
    color: #fff;
    font-family: Helvetica, Arial, Verdana, Tahoma, sans-serif;
    font-weight: 800;
    position: relative;
    z-index: 3;
    margin-left: 7%;
  }

  .header_top_search {
    display: block;
    float: right;
    width: 390px;
    margin-right: 80px;
    background: rgba(93, 155, 211, 0.01);
    position: relative;
    z-index: 3;

    .header_top_search_label {
      color: #fff;
      background-image: url(../../../images/magnifying-glass.png);
      width: 20px;
      height: 20px;
      background-position: center;
      background-size: contain;
      display: inline-block;
      position: absolute;
      right: 4%;
      top: 25%;
      cursor: pointer;
    }

    .header_top_search_input {
      width: 100%;
      height: 45px;
      border: 3.5px solid #fff;
      border-radius: 8%;
      position: relative;
      background: none;
      font-size: 18px;
      letter-spacing: 1px;
      line-height: 30px;
      color: #fff;
      font-family: Helvetica, Arial, Verdana, Tahoma, sans-serif;
      padding-left: 15px;
      padding-right: 15px;
      text-shadow: 0.5px 0.5px #000;
      display: inline-block;
    }

    .header_top_search_input::placeholder {
      color: #fff;
      opacity: 1;
    }

    .header_top_search_input:focus {
      outline: none;
    }
  }
}
