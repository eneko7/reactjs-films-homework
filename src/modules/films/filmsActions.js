import axios from 'axios';

export const FETCH_FILMS_REQUEST = 'FETCH_FILMS_REQUEST';
export const FETCH_FILMS_SUCCESS = 'FETCH_FILMS_SUCCESS';
export const FETCH_FILMS_ERROR = 'FETCH_FILMS_ERROR';
export const CLEAN_FILMS = 'CLEAN_FILMS';

export const urlPopularFilms = 'https://api.themoviedb.org/3/movie/popular?language=en-US&api_key=6a08c0def237c5910708279c9ee78cc5';
export const urlTopRatedFilms = 'https://api.themoviedb.org/3/movie/top_rated?language=en-US&api_key=6a08c0def237c5910708279c9ee78cc5';
export const urlComingSoonFilms = 'https://api.themoviedb.org/3/movie/upcoming?language=en-US&api_key=6a08c0def237c5910708279c9ee78cc5';
export const urlByGenreFilms = 'https://api.themoviedb.org/3/discover/movie?language=en-US&api_key=6a08c0def237c5910708279c9ee78cc5&sort_by=vote_average.desc&vote_count.gte=100';
export const urlBySearchFilms = 'https://api.themoviedb.org/3/search/movie?language=en-US&api_key=6a08c0def237c5910708279c9ee78cc5&include_adult=false';

export const receiveFilmsSuccess = payload => ({
  type: FETCH_FILMS_SUCCESS,
  payload,
});

export const receiveFilmsRequest = () => ({
  type: FETCH_FILMS_REQUEST,
});

export const receiveFilmsError = payload => ({
  type: FETCH_FILMS_ERROR,
  payload,
});

export const cleanFilms = () => ({
  type: CLEAN_FILMS,
});

export function fetchFilms(url, page = 1) {
  return (dispatch) => {
    dispatch(receiveFilmsRequest());
    axios.get(`${url}&page=${page}`)
      .then((response) => {
        dispatch(receiveFilmsSuccess({ films: response.data.results, url, page }));
      })
      .catch(error => dispatch(receiveFilmsError(error.message)));
  };
}

export function fetchFilmsPopular() {
  return (dispatch) => {
    dispatch(fetchFilms(urlPopularFilms));
  };
}

export function fetchFilmsTopRated() {
  return (dispatch) => {
    dispatch(fetchFilms(urlTopRatedFilms));
  };
}

export function fetchFilmsComingSoon() {
  return (dispatch) => {
    dispatch(fetchFilms(urlComingSoonFilms));
  };
}

export function fetchFilmsByGenre(genreID) {
  return (dispatch) => {
    dispatch(fetchFilms(`${urlByGenreFilms}&with_genres=${genreID}`));
  };
}

export function fetchFilmsBySearch(word) {
  return (dispatch) => {
    dispatch(fetchFilms(`${urlBySearchFilms}&query=${word}`));
  };
}

export function fetchNextFilms() {
  return (dispatch, getState) => {
    const state = getState();
    if (state.films.lastPage !== -1) {
      dispatch(fetchFilms(
        state.films.url,
        state.films.lastPage + 1,
      ));
    }
  };
}
