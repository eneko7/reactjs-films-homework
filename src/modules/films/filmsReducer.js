import {
  FETCH_FILMS_REQUEST,
  FETCH_FILMS_SUCCESS,
  FETCH_FILMS_ERROR,
} from './filmsActions';

const initialState = {
  allFilms: [],
  errorFilms: {},
  isFetchingFilms: false,
  isFetchedFilms: false,
  url: '',
  lastPage: null,
};

const filmsReducer = (state = initialState, action) => {
  switch (action.type) {
    case FETCH_FILMS_SUCCESS:
      return {
        ...state,
        isFetchingFilms: false,
        isFetchedFilms: true,
        allFilms: action.payload.page === 1
          ? action.payload.films
          : state.allFilms.concat(action.payload.films),
        url: action.payload.url,
        lastPage: action.payload.page,
        errorFilms: '',
      };
    case FETCH_FILMS_REQUEST:
      return {
        ...state,
        isFetchingFilms: true,
        isFetchedFilms: false,
      };
    case FETCH_FILMS_ERROR:
      return {
        ...state,
        errorFilms: action.payload,
        isFetchingFilms: false,
        isFetchedFilms: false,
        allFilms: [],
      };
    default:
      return {
        ...state,
      };
  }
};

export default filmsReducer;
