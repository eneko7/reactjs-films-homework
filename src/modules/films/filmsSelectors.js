export const getFilms = state => state.films.allFilms;
export const getIsFetchingFilms = state => state.films.isFetchingFilms;
export const getIsFetchedFilms = state => state.films.isFetchedFilms;
export const getErrorFilms = state => state.films.errorFilms;
