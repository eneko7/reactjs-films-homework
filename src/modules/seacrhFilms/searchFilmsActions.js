import {
  fetchFilmsBySearch,
} from '../films/filmsActions';

export const SAVE_SEARCHING_WORD = 'SAVE_SEARCHING_WORD';


export const saveSearchingWord = payload => (dispatch) => {
  dispatch({
    type: SAVE_SEARCHING_WORD,
    payload,
  });
};

export const searchFilm = payload => (dispatch) => {
  if (payload !== '') {
    dispatch(fetchFilmsBySearch(payload));
  }
};
