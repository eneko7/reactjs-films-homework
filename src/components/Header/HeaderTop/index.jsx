import HeaderTop from './HeaderTopContainer';

export default HeaderTop;
