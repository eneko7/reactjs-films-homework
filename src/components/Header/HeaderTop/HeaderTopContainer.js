import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import HeaderTop from './HeaderTop';
import { saveSearchingWord, searchFilm } from '../../../modules/seacrhFilms/searchFilmsActions';
import { fetchFilmsPopular } from '../../../modules/films/filmsActions';
import getSearchingWord from '../../../modules/seacrhFilms/searchFilmsSelectors';

const mapStateToProps = state => ({
  searchedWord: getSearchingWord(state),
});

const mapDispatchToProps = {
  saveSearchingWord,
  searchFilm,
  fetchFilmsPopular,
};
export default withRouter(connect(mapStateToProps, mapDispatchToProps)(HeaderTop));
