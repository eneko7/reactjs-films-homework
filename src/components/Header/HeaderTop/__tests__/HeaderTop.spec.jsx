import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import TestRenderer from 'react-test-renderer';
import moxios from 'moxios';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import * as searchFilmsActions from '../../../../modules/seacrhFilms/searchFilmsActions';
import * as filmsActions from '../../../../modules/films/filmsActions';
import HeaderTop from '../index';

const shallow = new ShallowRenderer();
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const store = mockStore({
  search: {
    searchedWord: 'word',
  },
});

describe('Search Snapshot', () => {
  test('snapshot render', () => {
    const component = shallow.render(
      <Provider store={store}>
        <HeaderTop />
      </Provider>,
    );
    expect(component).toMatchSnapshot();
  });
  test('dispatch save word', (done) => {
    const testRenderer = TestRenderer.create(
      <Provider store={store}>
        <HeaderTop />
      </Provider>,
    );
    moxios.stubRequest(/api.themoviedb.org/, {
      status: 200,
      response: '',
    });
    const e = {
      target: {
        value: 'word',
      },
    };

    const expectedActions = [
      { type: searchFilmsActions.SAVE_SEARCHING_WORD, payload: 'word' },
      { type: filmsActions.FETCH_FILMS_REQUEST },
    ];

    const { root } = testRenderer;
    root.findByProps({ className: 'header_top_search_input' }).props.onChange(e);
    moxios.wait(() => {
      expect(store.getActions()).toEqual(expectedActions);
      done();
    });
  });
});
