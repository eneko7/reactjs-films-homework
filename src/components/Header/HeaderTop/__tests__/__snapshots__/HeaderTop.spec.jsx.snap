// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Snapshot snapshot render 1`] = `<Connect(HeaderTop) />`;
